subsys/net/l2/openthread/openthread_border_router.c

```diff
@@ -0,0 +1,185 @@
+
+static struct net_mgmt_event_callback ail_net_event_connection_cb;
+static struct net_icmp_ctx ra_ctx;
+static struct net_icmp_ctx rs_ctx;
+static uint32_t ail_iface_index;
+static struct net_if *ail_iface_ptr;
+static bool is_border_router_started;
+
+int openthread_start_border_router_services(struct net_if *ot_iface, struct net_if *ail_iface)
+{
+	otError error = OT_ERROR_NONE;
+	otInstance *instance = openthread_get_default_instance();
+
+	ail_iface_index = (uint32_t)net_if_get_by_iface(ail_iface);
+	ail_iface_ptr = ail_iface;
+
+	net_if_flag_set(ot_iface, NET_IF_FORWARD_MULTICASTS);
+
+	openthread_mutex_lock();
+	/* Initialize platform modules first */
+	VerifyOrExit(infra_if_init(instance, ail_iface_ptr) == OT_ERROR_NONE,
+		     error = OT_ERROR_FAILED);
+
+	/* Call OpenThread API */
+	VerifyOrExit(otBorderRoutingInit(instance, ail_iface_index, true) == OT_ERROR_NONE,
+		     error = OT_ERROR_FAILED);
+	VerifyOrExit(otBorderRoutingSetEnabled(instance, true) == OT_ERROR_NONE,
+		     error = OT_ERROR_FAILED);
+	VerifyOrExit(otPlatInfraIfStateChanged(instance, ail_iface_index, true) == OT_ERROR_NONE,
+		     error = OT_ERROR_FAILED);
+	otBackboneRouterSetEnabled(instance, true);
+	is_border_router_started = true;
+
+exit:
+	openthread_mutex_unlock();
+	return error == OT_ERROR_NONE ? 0 : -EIO;
+}
+
+static int openthread_stop_border_router_services(struct net_if *ot_iface, struct net_if *ail_iface)
+{
+	otError error = OT_ERROR_FAILED;
+	otInstance *instance = openthread_get_default_instance();
+
+	openthread_mutex_lock();
+
+	if (is_border_router_started) {
+		/* Call OpenThread API */
+		VerifyOrExit(otPlatInfraIfStateChanged(instance, ail_iface_index, false) ==
+				     OT_ERROR_NONE,
+			     error = OT_ERROR_FAILED);
+		VerifyOrExit(otBorderRoutingSetEnabled(instance, false) == OT_ERROR_NONE,
+			     error = OT_ERROR_FAILED);
+		otBackboneRouterSetEnabled(instance, false);
+
+		error = OT_ERROR_NONE;
+		is_border_router_started = false;
+	}
+exit:
+	openthread_mutex_unlock();
+	return error == OT_ERROR_NONE ? 0 : -EIO;
+}
+
+void openthread_set_bbr_multicast_listener_cb(openthread_bbr_multicast_listener_cb cb,
+					      void *context)
+{
+	__ASSERT(cb != NULL, "Receive callback is not set");
+	__ASSERT(openthread_instance != NULL, "OpenThread instance is not initialized");
+
+	openthread_mutex_lock();
+	otBackboneRouterSetMulticastListenerCallback(openthread_get_default_instance(), cb,
+						     context);
+	openthread_mutex_unlock();
+}
+
+static int handle_ra_input(struct net_icmp_ctx *ctx, struct net_pkt *pkt,
+			   struct net_icmp_ip_hdr *hdr, struct net_icmp_hdr *icmp_hdr,
+			   void *user_data)
+{
+	otInstance *ot_instance = openthread_get_default_instance();
+	otIp6Address src_addr = {0};
+	uint16_t length = net_pkt_get_len(pkt);
+	uint8_t payload[512] = {0};
+	uint8_t *p = payload;
+
+	if (net_buf_linearize(p, sizeof(payload), pkt->buffer, 0, pkt->buffer->len) ==
+	    pkt->buffer->len) {
+		p = &payload[0];
```

Comment:
`p` is already set to point to `payload` 4 lines above. Not sure why you even need it. You could use `&payload[sizeof(struct net_ipv6_hdr)]` directly.